### PR TITLE
PYIC-5403: Convert BuildProvenUserIdentityDetails to an API GW proxy

### PR DESCRIPTION
**Pull request by Joe-Edwards-GDS**
Previously this endpoint used VTL to transform the request/response, but a proxy passthrough removes the need to use VTL, and keeps all processing logic in the lambda code.

### CODE REVIEW

File: lambdas/build-proven-user-identity-details/src/main/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/BuildProvenUserIdentityDetailsHandler.java [hunk 1]
```diff
@@ -140,11 +140,14 @@ public Map<String, Object> handleRequest(JourneyRequest input, Context context)
...
+        return buildJourneyErrorResponse(errorResponse, HttpStatus.SC_INTERNAL_SERVER_ERROR);
+    }
+
+    private APIGatewayProxyResponseEvent buildJourneyErrorResponse(
+            ErrorResponse errorResponse, int statusCode) {
         LOGGER.error(LogHelper.buildLogMessage(errorResponse.getMessage()));
-        return new JourneyErrorResponse(
-                        JOURNEY_ERROR_PATH, HttpStatus.SC_INTERNAL_SERVER_ERROR, errorResponse)
-                .toObjectMap();
+        return ApiGatewayResponseGenerator.proxyJsonResponse(statusCode, errorResponse);
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-proven-user-identity-details/src/main/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/BuildProvenUserIdentityDetailsHandler.java:150]: This is a small change, so error responses from this endpoint no longer look like journey events.

old:
```
{
  "journey": "/journey/error",
  "statusCode": 500,
  "code": 1234,
  "message": "an error happened"
}
```

new:
```
{
  "code": 1234,
  "message": "an error happened"
}
```
core-front doesn't do anything with this anyway, except include it in the logs

File: lambdas/build-proven-user-identity-details/src/main/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/domain/ProvenUserIdentityDetails.java [hunk 2]
```diff
@@ -7,15 +7,14 @@
...
 import uk.gov.di.ipv.core.library.domain.Address;
-import uk.gov.di.ipv.core.library.domain.BaseResponse;
 
 import java.util.List;
 
 @EqualsAndHashCode(callSuper = false)
 @ExcludeFromGeneratedCoverageReport
 @Data
 @Builder
-public class ProvenUserIdentityDetails extends BaseResponse {
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-proven-user-identity-details/src/main/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/domain/ProvenUserIdentityDetails.java:18]: (`BaseResponse` looks a bit suspect in general, but we don't need it here - we can serialize this class normally)

File: lambdas/build-proven-user-identity-details/src/test/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/BuildProvenUserIdentityDetailsHandlerTest.java [hunk 3]
```diff
@@ -121,7 +119,7 @@ void shouldReceive200ResponseCodeProvenUserIdentityDetails() throws Exception {
         when(mockClientOAuthSessionDetailsService.getClientOAuthSession(any()))
                 .thenReturn(clientOAuthSessionItem);
 
-        JourneyRequest input = createRequestEvent();
+        var input = createRequestEvent();
```
Comment (Joe-Edwards-GDS) [inline lambdas/build-proven-user-identity-details/src/test/java/uk/gov/di/ipv/core/buildprovenuseridentitydetails/BuildProvenUserIdentityDetailsHandlerTest.java:122]: (a good advertisement for using `var`, I think)

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/ErrorResponse.java [hunk 4]
```diff
@@ -82,33 +86,11 @@ public enum ErrorResponse {
...
-        return String.format("{\"code\":%d,\"message\":\"%s\"}", code, message);
-    }
-
-    public String getMessage() {
-        return message;
-    }
-
     @JsonCreator
-    public static ErrorResponse forCode(int code) {
+    public static ErrorResponse forCode(@JsonProperty("code") int code) {
```
Comment (Joe-Edwards-GDS) [inline libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/ErrorResponse.java:93]: This makes the serialisation and deserialisation for this class symmetric - previously it serialised to an object, but deserialised from an int (the annotation here turns this into a 'properties' JsonCreator, which reads the property from a JSON object).

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/JourneyErrorResponse.java
Comment (Joe-Edwards-GDS) [inline libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/JourneyErrorResponse.java:1]: The JSON properties on this class are a bit of a mess, but it needed a small change as previously it was relying on the asymmetric serialisation of `ErrorResponse` (now it only ever serialises/deserialises using ints).

File: local-running/src/main/java/uk/gov/di/ipv/coreback/handlers/LambdaHandler.java [hunk 5]
```diff
@@ -144,39 +132,28 @@ public LambdaHandler() throws IOException {
...
-        APIGatewayProxyRequestEvent apiGatewayProxyRequestEvent = new APIGatewayProxyRequestEvent();
-        apiGatewayProxyRequestEvent.setBody(request.body());
-        apiGatewayProxyRequestEvent.setHeaders(getHeadersMap(request));
-
-        APIGatewayProxyResponseEvent responseEvent =
-                handler.handleRequest(apiGatewayProxyRequestEvent, EMPTY_CONTEXT);
-
-        response.type(APPLICATION_JSON);
-        return responseEvent.getBody();
+        return (Request request, Response response) -> {
```
Comment (Wynndow) [inline local-running/src/main/java/uk/gov/di/ipv/coreback/handlers/LambdaHandler.java:145]: nice

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/ErrorResponse.java (hunk 4 above)
Comment (Wynndow) [reply libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/ErrorResponse.java:93]: I've actually had a long standing desire to remove this ErrorResponse class completely. I'm not sure it actually does anything for us.

**Review by Wynndow — APPROVED** [review]

File: libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/ErrorResponse.java (hunk 4 above)
Comment (Joe-Edwards-GDS) [reply libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/ErrorResponse.java:93]: Having a standard list of error codes is potentially useful, but I agree we're not really making any use of it!

In principle, these errors should be monitored and visible in logs etc., with a corresponding run-book entry for each (i.e. what to do if each of the errors occurs).

(If you're not doing that, you may as well just inline the error messages...)

**Pull request merged**